codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java

```diff
@@ -254,6 +259,82 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws
+                    // out because we started without a pending #bufferPromise.
+                    return combinedPromise(ctx, promise, ctx.newPromise());
+                }
+                bufferPromise = promise;
+                // out is empty and we just return a new promise as out does not contain any part of msg
+                return ctx.newPromise();
+            }
+            return promise;
+        }
+        if (buffer.isReadable()) {
+            if (out.isReadable()) {
+                // the second half of msg got buffered in #buffer and the first half went into out so both have to
+                // succeed for msg's promise to succeed so we associate #bufferPromise with out
+                return combinedPromise(ctx, promise, bufferPromise);
+            }
+            bufferPromise.addListener(new ChannelPromiseNotifier(promise));
+            // out is empty and we just return a new promise as out does not contain any part of msg
+            return ctx.newPromise();
+        }
+        // msg exactly filled up a frame so both the previously buffered data as well as msg are now in out so we
+        // notify both simultaneously through #bufferPromise
+        bufferPromise.addListener(new ChannelPromiseNotifier(promise));
+        final ChannelPromise prev = bufferPromise;
+        bufferPromise = null;
+        return prev;
+    }
+
+    /**
+     * This method handles the promise ({@code msgPromise}) to notify for a message that is split between the current
+     * output buffer and {@link #buffer}. In this case {@code msgPromise} is notified with success once the current
+     * {@link #bufferPromise} succeeds and the promise ({@code outPromise}) associated with the current write buffer
+     * succeeds. If either of the two do not succeed then the {@code msgPromise} must fail.
+     *
+     * @param ctx        the {@link ChannelHandlerContext} which this {@link Lz4FrameEncoder} belongs to
+     * @param msgPromise promise to notify once all of a message has been written
+     * @param outPromise promise to notify once the current write buffer has been written
+     * @return           the given {@code outPromise}
+     */
+    private ChannelPromise combinedPromise(final ChannelHandlerContext ctx, final ChannelPromise msgPromise,
+                                           final ChannelPromise outPromise) {
+        final ChannelPromise outstandingMsgPartPromise = ctx.newPromise();
+        bufferPromise = outstandingMsgPartPromise;
+        outPromise.addListener(new ChannelFutureListener() {
+            @Override
+            public void operationComplete(ChannelFuture future) throws Exception {
+                if (future.isSuccess()) {
+                    outstandingMsgPartPromise.addListener(new ChannelPromiseNotifier(msgPromise));
+                } else {
+                    msgPromise.setFailure(future.cause());
```

Comment:
Don't we have to do anything with the `outstandingMsgPartPromise` (`bufferPromise`) here? If not, where would we complete that `promise`?

Reply:
See https://github.com/netty/netty/pull/11419#discussion_r663922122, this should be clearer now that the final promise depends on both parts in all cases I hope.